### PR TITLE
controller: handle read/write beyond volume condition inside the cont…

### PR DESCRIPTION
**Pull request by kp6**
…roller to prevent replicas from being kicked out

- issue #131

### CODE REVIEW

File: controller/control.go [hunk 1]
```diff
@@ -234,6 +234,11 @@ func (c *Controller) Start(addresses ...string) error {
 
 func (c *Controller) WriteAt(b []byte, off int64) (int, error) {
 	c.RLock()
+	if off < 0 || off+int64(len(b)) > c.size {
```
Comment (yasker) [inline controller/control.go]: Should be `>=`...

**Comment by yasker** [comment]
We should also enable the integration test for this. In the same commit.

File: controller/control.go (hunk 1 above)
Comment (kp6) [inline controller/control.go]: yep

Comment (kp6) [inline controller/control.go]: hold on, no, why `>=`? 
Reading 100 bytes at offset 0 is fine for a 100 byte volume

Comment (yasker) [inline controller/control.go]: Oh yeah, you're right...

Comment (yasker) [inline controller/control.go]: Btw, we can add this case to boundary test case as well.

Comment (kp6) [inline controller/control.go]: ok, what's the simplest way to launch just one selected integration test?

Comment (yasker) [inline controller/control.go]: `./scripts/integration-test -- -k test_beyond_bounary`(found I wrote a typo here)

After `make build`, you need to refresh you local binary for this to run correctly.

Comment (kp6) [inline controller/control.go]: could you elaborate?
refresh what binary? `bin/longhorn`? or which? what do you mean by refresh?

Comment (yasker) [inline controller/control.go]: `bin/longhorn` I meant.

if `make` failed, you won't get new `bin/longhorn` binary with your latest change. So you may need to run `make build` to get binary, and skip all the tests to get the latest binary.

Comment (yasker) [inline controller/control.go]: Because the command I provided above is running in the host, rather inside container.

Comment (kp6) [inline controller/control.go]: I see... so I can only ran it on rancher os then?

Comment (yasker) [inline controller/control.go]: ... Why? How you normally build?

Comment (kp6) [inline controller/control.go]: I build on ubuntu: `make` or `make build`

Comment (kp6) [inline controller/control.go]: here's what I get:
```
kirill@ubuntu:~/go/src/github.com/rancher/longhorn$ ./scripts/integration-test -- -k test_beyond_bounary
longhorn: no process found
ssync: no process found
INFO[0000] Listening on localhost:9501                  
INFO[0000] Listening on control localhost:9505          
INFO[0000] Listening on data localhost:9506             
INFO[0000] Listening on data localhost:9503             
INFO[0000] Listening on control localhost:9502          
INFO[0000] Listening on sync localhost:9507             
INFO[0000] Listening on sync localhost:9504             
./scripts/integration-test: line 28: qemu-img: command not found

Comment (yasker) [inline controller/control.go]: Yeah, so why you need to run it on RancherOS?

Comment (yasker) [inline controller/control.go]: Just install the missing tools...

Comment (kp6) [inline controller/control.go]: ok, but what does this mean:
```
longhorn: no process found
ssync: no process found

Comment (kp6) [inline controller/control.go]: the test seems to run fine despite the above though....

**Comment by kp6** [comment]
enabled slightly amended test

File: integration/data/test_basic_ops.py [hunk 2]
```diff
@@ -16,16 +16,20 @@ def test_basic_rw(dev):  # NOQA
         common.verify_data(dev, offset, data)
 
 
-# TODO BUG: https://github.com/rancher/longhorn/issues/131
```
Comment (yasker) [inline integration/data/test_basic_ops.py]: Better to keep the bug here for future reference.

**Comment by kp6** [comment]
amended

**Pull request merged**